### PR TITLE
Remove Jpeg Huffman Decoder Bottleneck

### LINKED ISSUE #601: Faster Jpeg Huffman Scan Decoding
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of ImageSharp
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

With #571 now included in the codebase we have an opportunity to further enhance the performance of our jpeg decoder.

There is one obvious candidate for improvement within the decoder: The decoding of the SOS (Start of Scan) segment. 

Our approach works but is naïve. We only read/decode one byte at a time and do not use optimized tables for Huffman code lookup. There are further optimizations available with AC Huffman decoding that we also do not do. 

I believe that with the correct implementation we can reduce our decoding time by at least 10ms when compared to our current benchmarks.

### Inspiration for better code
- [MozJpeg - (LibJpeg-Turbo port)](https://github.com/mozilla/mozjpeg/blob/master/jchuff.c)
- StbImage - (https://github.com/nothings/stb/blob/e6afb9cbae4064da8c3e69af3ff5c4629579c1d2/stb_image.h#L1680)
- [Rust Jpeg Decoder](https://github.com/kaksmet/jpeg-decoder/blob/35c080d7dca452fbec534f5ac22664e1cc3c1a1c/src/huffman.rs)

I'm looking for help here. I've made several attempts to port from both sources but have failed miserably so far.

**Update**

@bitbank2 Has kindly written some information with pointers to demonstrate how to optimise entropy decoding.

http://bitbanksoftware.blogspot.com/2018/06/optimizing-jpeg-entropy-decoding.html

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Replaces the slow PdfJsScanDecoder implementation with a much faster implementation based on Stb_Image. Now it's no longer a bottleneck. Fixes #601 

We're creeping slowly closer to System.Drawing now for jpeg decoding.

**New Benchmarks**
``` console
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
Frequency=2742191 Hz, Resolution=364.6719 ns, Timer=TSC
.NET Core SDK=2.1.300
  [Host]     : .NET Core 2.0.7 (CoreCLR 4.6.26328.01, CoreFX 4.6.26403.03), 64bit RyuJIT
  Job-CBPIPJ : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3110.0
  Job-FVNPWQ : .NET Core 2.0.7 (CoreCLR 4.6.26328.01, CoreFX 4.6.26403.03), 64bit RyuJIT

LaunchCount=1  TargetCount=3  WarmupCount=3

                           Method | Runtime |                    TestImage |      Mean |     Error |    StdDev | Scaled | ScaledSD |    Gen 0 | Allocated |
--------------------------------- |-------- |----------------------------- |----------:|----------:|----------:|-------:|---------:|---------:|----------:|
   'Decode Jpeg - System.Drawing' |     Clr |  Jpg/baseline/Calliphora.jpg |  6.786 ms |  2.364 ms | 0.1336 ms |   1.00 |     0.00 | 117.1875 | 254.47 KB |
       'Decode Jpeg - ImageSharp' |     Clr |  Jpg/baseline/Calliphora.jpg | 33.662 ms |  4.877 ms | 0.2755 ms |   4.96 |     0.09 |        - |  52.63 KB |
 'Decode Jpeg - ImageSharp PdfJs' |     Clr |  Jpg/baseline/Calliphora.jpg | 24.767 ms |  3.341 ms | 0.1888 ms |   3.65 |     0.06 |        - |  25.25 KB |
                                  |         |                              |           |           |           |        |          |          |           |
   'Decode Jpeg - System.Drawing' |    Core |  Jpg/baseline/Calliphora.jpg |  7.079 ms | 11.763 ms | 0.6646 ms |   1.00 |     0.00 | 117.1875 | 254.11 KB |
       'Decode Jpeg - ImageSharp' |    Core |  Jpg/baseline/Calliphora.jpg | 34.337 ms | 12.214 ms | 0.6901 ms |   4.88 |     0.36 |        - |  47.79 KB |
 'Decode Jpeg - ImageSharp PdfJs' |    Core |  Jpg/baseline/Calliphora.jpg | 26.188 ms |  1.067 ms | 0.0603 ms |   3.72 |     0.27 |        - |  21.71 KB |
                                  |         |                              |           |           |           |        |          |          |           |
   'Decode Jpeg - System.Drawing' |     Clr | Jpg/baseline/jpeg420exif.jpg | 20.881 ms | 21.458 ms | 1.2124 ms |   1.00 |     0.00 | 343.7500 | 757.89 KB |
       'Decode Jpeg - ImageSharp' |     Clr | Jpg/baseline/jpeg420exif.jpg | 80.619 ms |  6.260 ms | 0.3537 ms |   3.87 |     0.18 | 250.0000 | 564.28 KB |
 'Decode Jpeg - ImageSharp PdfJs' |     Clr | Jpg/baseline/jpeg420exif.jpg | 56.485 ms |  7.069 ms | 0.3994 ms |   2.71 |     0.13 | 250.0000 | 535.07 KB |
                                  |         |                              |           |           |           |        |          |          |           |
   'Decode Jpeg - System.Drawing' |    Core | Jpg/baseline/jpeg420exif.jpg | 17.653 ms |  4.603 ms | 0.2601 ms |   1.00 |     0.00 | 343.7500 | 757.04 KB |
       'Decode Jpeg - ImageSharp' |    Core | Jpg/baseline/jpeg420exif.jpg | 80.280 ms | 10.145 ms | 0.5732 ms |   4.55 |     0.06 | 250.0000 | 548.49 KB |
 'Decode Jpeg - ImageSharp PdfJs' |    Core | Jpg/baseline/jpeg420exif.jpg | 57.506 ms |  6.344 ms | 0.3584 ms |   3.26 |     0.04 | 250.0000 | 522.19 KB |
```
**Before Performance Trace**
Before Performance tracing details

**After Performance Trace**
After Performance tracing details

I'm genuinely amazed I managed to pull this off!

### CODE REVIEW

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FastACTables.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+using System;
+using SixLabors.Memory;
+
+namespace SixLabors.ImageSharp.Formats.Jpeg.PdfJsPort.Components
+{
+    /// <summary>
+    /// The collection of lookup tables used for fast AC entropy scan decoding.
+    /// </summary>
+    internal sealed class FastACTables : IDisposable
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FastACTables.cs]: I wonder if we really need this class in it's current form. Either we should hide the `Table` property and encapsulate actual logic (initialization, retrieval of rows) inside the class, or we should drop it and define a helper method that creates  `Buffer2d<short>`.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+        private readonly int successiveLow;
+
+        private int codeBits;
+        private uint codeBuffer;
+        private bool nomore;
+        private bool eof;
+        private byte marker;
+        private bool badMarker;
+        private long markerPosition;
+        private int todo;
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: Would be nice to have better names *or* some comments about the semantics of less trivial variables.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+                {
+                    for (int i = 0; i < mcusPerLine; i++)
+                    {
+                        // Scan an interleaved mcu... process components in order
+                        for (int k = 0; k < this.componentsLength; k++)
+                        {
+                            PdfJsFrameComponent component = this.components[k];
+                            ref short blockDataRef = ref MemoryMarshal.GetReference(MemoryMarshal.Cast<Block8x8, short>(component.SpectralBlocks.Span));
+                            ref PdfJsHuffmanTable dcHuffmanTable = ref dcHuffmanTables[component.DCHuffmanTableId];
+                            ref PdfJsHuffmanTable acHuffmanTable = ref acHuffmanTables[component.ACHuffmanTableId];
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: `acHuffmanTable` and `fastAC`  are unused

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+            {
+                // Refinement scan for DC coefficient
+                if (this.GetBit() != 0)
+                {
+                    blockDataRef += (short)(1 << this.successiveLow);
+                }
+            }
+        }
+
+        private void DecodeBlockProgressiveAC(
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: The method is quite long, might be worth to refactor internal branches into separate methods.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+            {
+                this.GrowBufferUnsafe();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int PeekBits() => (int)((this.codeBuffer >> (32 - FastBits)) & ((1 << FastBits) - 1));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private uint LRot(uint x, int y) => (x << y) | (x >> (32 - y));
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: The method can be static

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FixedInt64Buffer18.cs [hunk 6]
```diff
@@ -9,14 +9,14 @@ namespace SixLabors.ImageSharp.Formats.Jpeg.PdfJsPort.Components
     [StructLayout(LayoutKind.Sequential)]
     internal unsafe struct FixedInt64Buffer18
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FixedInt64Buffer18.cs]: Good catch!

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FastACTables.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FastACTables.cs]: I like having a class but I think you're tight about the tables property. I'll hide that away and introduce an indexer.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: Will do!

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: Sometimes I miss Resharper.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs (hunk 4 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: I'll split out the Refinement part into a separate method.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs (hunk 5 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: Good catch!

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs]: Got the unused, different line though so Github won't see it.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/PdfJsJpegDecoderCore.cs [hunk 7]
```diff
@@ -837,5 +857,47 @@ private Image<TPixel> PostProcessIntoImage<TPixel>()
                 return image;
             }
         }
+
+        /// <summary>
+        /// Builds a lookup table for fast AC entropy scan decoding.
+        /// </summary>
+        /// <param name="index">The table index.</param>
+        private void BuildFastACTable(int index)
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/PdfJsJpegDecoderCore.cs]: This should be probably a member of the `FastACTables` class.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FixedInt16Buffer18.cs [hunk 8]
```diff
@@ -9,14 +9,14 @@ namespace SixLabors.ImageSharp.Formats.Jpeg.PdfJsPort.Components
     [StructLayout(LayoutKind.Sequential)]
     internal unsafe struct FixedInt16Buffer18
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/FixedInt16Buffer18.cs]: Should we rename the struct because it no longer is a `short`?

**Comment by antonfirsov** [comment]
@JimBobSquarePants pushed some code simplification, and a benchmark to quickly test `ParseStream()`. There is no change in performance:

#### Baseline, Before:
```
                           Method | Runtime |                    TestImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |    Gen 0 | Allocated |
--------------------------------- |-------- |----------------------------- |---------:|----------:|----------:|-------:|---------:|---------:|----------:|
            'System.Drawing FULL' |     Clr | Jpg/baseline/jpeg420exif.jpg | 18.35 ms |  8.224 ms | 0.4647 ms |   1.00 |     0.00 | 218.7500 | 757.88 KB |
 PdfJsJpegDecoderCore.ParseStream |     Clr | Jpg/baseline/jpeg420exif.jpg | 19.64 ms | 13.984 ms | 0.7901 ms |   1.07 |     0.04 |        - |     15 KB |
                                  |         |                              |          |           |           |        |          |          |           |
            'System.Drawing FULL' |    Core | Jpg/baseline/jpeg420exif.jpg | 18.36 ms |  4.017 ms | 0.2270 ms |   1.00 |     0.00 | 218.7500 | 757.04 KB |
 PdfJsJpegDecoderCore.ParseStream |    Core | Jpg/baseline/jpeg420exif.jpg | 21.21 ms | 10.237 ms | 0.5784 ms |   1.16 |     0.03 |        - |  14.84 KB |

```

#### Baseline, After:
```
                           Method | Runtime |                    TestImage |     Mean |    Error |    StdDev | Scaled | ScaledSD |    Gen 0 | Allocated |
--------------------------------- |-------- |----------------------------- |---------:|---------:|----------:|-------:|---------:|---------:|----------:|
            'System.Drawing FULL' |     Clr | Jpg/baseline/jpeg420exif.jpg | 18.38 ms | 7.007 ms | 0.3959 ms |   1.00 |     0.00 | 218.7500 | 757.88 KB |
 PdfJsJpegDecoderCore.ParseStream |     Clr | Jpg/baseline/jpeg420exif.jpg | 19.71 ms | 1.502 ms | 0.0849 ms |   1.07 |     0.02 |        - |     15 KB |
                                  |         |                              |          |          |           |        |          |          |           |
            'System.Drawing FULL' |    Core | Jpg/baseline/jpeg420exif.jpg | 17.86 ms | 1.077 ms | 0.0609 ms |   1.00 |     0.00 | 218.7500 | 757.04 KB |
 PdfJsJpegDecoderCore.ParseStream |    Core | Jpg/baseline/jpeg420exif.jpg | 20.84 ms | 7.756 ms | 0.4382 ms |   1.17 |     0.02 |        - |  14.88 KB |
```

**Comment by JimBobSquarePants** [comment]
Nice changes! That class looks very clean now and I like the inline and error helpers!

How the hell does S.D decode the full stream so damn quickly, it's infuriating!

**Comment by antonfirsov** [comment]
Don't ask how, but it's possible to optimize huffman decoding with SIMD.

**Comment by JimBobSquarePants** [comment]
I'm gonna merge this now to work on namespacing.

**Comment by welkie** [comment]
Came here after reading the recent release notes. I notice you say it's moving closer to `System.Drawing` for JPEG decoding. I was reading online and in the MSDN docs, came across [this page](https://docs.microsoft.com/en-us/dotnet/api/system.drawing?view=netstandard-2.0), saying this:

> Classes within the System.Drawing namespace are not supported for use within a Windows or ASP.NET service. Attempting to use these classes from within one of these application types may produce unexpected problems, such as diminished service performance and run-time exceptions. For a supported alternative, see Windows Imaging Components.

Is it a good idea to use ImageSharp for image processing in an ASP.NET Core web app?

**Pull request merged**